lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -2124,6 +2144,47 @@ void MemoryToRegisters::removeSingleBlockAllocation(AllocStackInst *asi) {
   }
 }
 
+void MemoryToRegisters::collectStoredValues(AllocStackInst *asi,
+                                            StackList<SILValue> &owned,
+                                            StackList<SILValue> &guaranteed) {
+  for (auto *use : asi->getUses()) {
+    auto *user = use->getUser();
+    if (auto *si = dyn_cast<StoreInst>(user)) {
+      owned.push_back(si->getSrc());
+    } else if (auto *sbi = dyn_cast<StoreBorrowInst>(user)) {
+      guaranteed.push_back(sbi->getSrc());
+    }
+  }
+}
+
+void MemoryToRegisters::canonicalizeValueLifetimes(
+    StackList<SILValue> &owned, StackList<SILValue> &guaranteed) {
+  CanonicalizeOSSALifetime canonicalizer(
+      /*pruneDebug=*/true, /*maximizeLifetime=*/!f.shouldOptimize(), &f,
+      accessBlockAnalysis, domInfo, calleeAnalysis, deleter);
+  for (auto value : owned) {
+    auto root = CanonicalizeOSSALifetime::getCanonicalCopiedDef(value);
+    if (auto *copy = dyn_cast<CopyValueInst>(root)) {
+      if (SILValue borrowDef = CanonicalizeBorrowScope::getCanonicalBorrowedDef(
+              copy->getOperand())) {
+        guaranteed.push_back(copy);
+        continue;
+      }
+    }
+    canonicalizer.canonicalizeValueLifetime(root);
+  }
+  CanonicalizeBorrowScope borrowCanonicalizer(&f, deleter);
+  for (auto value : guaranteed) {
+    auto borrowee = CanonicalizeBorrowScope::getCanonicalBorrowedDef(value);
+    if (!borrowee)
+      continue;
+    BorrowedValue borrow(borrowee);
+    if (borrow.kind != BorrowedValueKind::SILFunctionArgument)
+      continue;
+    borrowCanonicalizer.canonicalizeBorrowScope(borrow);
```

Comment:
Just curious, is this the first place where we actually canonicalize borrows ? I believe its turned off in the copy propagation pass ?

Reply:
Though we could turn it on now https://github.com/apple/swift/pull/65687 thanks to some recent fixes, ( https://github.com/apple/swift/pull/65686 , https://github.com/apple/swift/pull/65458 ), we shouldn't because there are some regressions when doing so.  I am planning to look into those regressions and get it enabled before too long.

Actually, though, guaranteed lifetime canonicalization is already enabled for guaranteed function arguments in CopyPropagation and SILCombine, which is all that is being handled here for now.  

At some point, it might be nice to have a function that could be called from here or SILCombine which deals with the necessary post-processing.

Reply:
thanks!